core/src/main/java/com/sequenceiq/cloudbreak/converter/stack/cluster/gateway/GatewayJsonToGatewayConverter.java

```diff
@@ -1,17 +1,106 @@
+import com.sequenceiq.cloudbreak.api.model.stack.cluster.gateway.SSOType;
+import com.sequenceiq.cloudbreak.controller.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.controller.validation.ValidationResult;
+import com.sequenceiq.cloudbreak.controller.validation.ValidationResult.State;
+import com.sequenceiq.cloudbreak.controller.validation.stack.cluster.gateway.ExposedServiceListValidator;
 import com.sequenceiq.cloudbreak.converter.AbstractConversionServiceAwareConverter;
+import com.sequenceiq.cloudbreak.domain.json.Json;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.gateway.ExposedServices;
 import com.sequenceiq.cloudbreak.domain.stack.cluster.gateway.Gateway;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.gateway.GatewayTopology;
 
 @Component
 public class GatewayJsonToGatewayConverter extends AbstractConversionServiceAwareConverter<GatewayJson, Gateway> {
 
+    private final ExposedServiceListValidator exposedServiceListValidator;
+
+    public GatewayJsonToGatewayConverter(ExposedServiceListValidator exposedServiceListValidator) {
+        this.exposedServiceListValidator = exposedServiceListValidator;
+    }
+
     @Override
     public Gateway convert(GatewayJson source) {
+        Gateway gateway = new Gateway();
+        if (source.getGatewayType() != null) {
+            gateway.setGatewayType(source.getGatewayType());
+        }
+        if (source.getSsoType() != null) {
+            gateway.setSsoType(source.getSsoType());
+        } else {
+            gateway.setSsoType(SSOType.NONE);
+        }
+        gateway.setTokenCert(source.getTokenCert());
+        gateway = convertTopologies(source, gateway);
+        return gateway;
+    }
+
+    private Gateway convertTopologies(GatewayJson gatewayJson, Gateway gateway) {
+        if (isLegacyGatewayRequest(gatewayJson)) {
+            return convertLegacyTopology(gatewayJson, gateway);
+        } else {
+            return convertTopologyList(gatewayJson, gateway);
+        }
+    }
+
+    private boolean isLegacyGatewayRequest(GatewayJson source) {
+        return StringUtils.isNotBlank(source.getTopologyName());
+    }
+
+    private Gateway convertLegacyTopology(GatewayJson gatewayJson, Gateway gateway) {
+        try {
+            GatewayTopology gatewayTopology = new GatewayTopology();
+            gatewayTopology.setTopologyName(gatewayJson.getTopologyName());
+            gatewayTopology = convertLegacyExposedServices(gatewayJson, gatewayTopology);
+            gateway.setTopologies(Sets.newHashSet(gatewayTopology));
+            gatewayTopology.setGateway(gateway);
+            return gateway;
+        } catch (JsonProcessingException e) {
+            throw new BadRequestException("Invalid exposedServices in request. Could not be parsed to JSON.", e);
+        }
+    }
+
+    private GatewayTopology convertLegacyExposedServices(GatewayJson gatewayJson, GatewayTopology gatewayTopology) throws JsonProcessingException {
```

Comment:
Doesn't convert, it sets

Reply:
actually it uses the conversionService to convert the topologies